coderd/database/queries/prebuilds.sql

```diff
@@ -0,0 +1,141 @@
+WHERE wlb.job_status IN ('pending'::provisioner_job_status, 'running'::provisioner_job_status)
+GROUP BY t.id, wpb.template_version_id, wpb.transition;
+
+-- GetPresetsBackoff groups workspace builds by template version ID.
+-- For each group, the query checks up to N of the most recent jobs that occurred within the
+-- lookback period, where N equals the number of desired instances for the corresponding preset.
+-- If at least one of the job within a group has failed, we should backoff on the corresponding template version ID.
+-- Query returns a list of template version IDs for which we should backoff.
+-- Only active template versions with configured presets are considered.
+-- We also return the number of failed workspace builds that occurred during the lookback period.
+--
+-- NOTE:
+-- - To **decide whether to back off**, we look at up to the N most recent builds (within the defined lookback period).
+-- - To **calculate the number of failed builds**, we consider all builds within the defined lookback period.
+--
+-- The number of failed builds is used downstream to determine the backoff duration.
+-- name: GetPresetsBackoff :many
+WITH filtered_builds AS (
+	-- Only select builds which are for prebuild creations
+	SELECT wlb.template_version_id, wlb.created_at, tvp.id AS preset_id, wlb.job_status, tvp.desired_instances
+	FROM template_version_presets tvp
+			 INNER JOIN workspace_latest_builds wlb ON wlb.template_version_preset_id = tvp.id
+             INNER JOIN template_versions tv ON wlb.template_version_id = tv.id
+             INNER JOIN templates t ON tv.template_id = t.id AND t.active_version_id = tv.id
+	WHERE tvp.desired_instances IS NOT NULL -- Consider only presets that have a prebuild configuration.
+      AND wlb.transition = 'start'::workspace_transition
+),
+time_sorted_builds AS (
+    -- Group builds by template version, then sort each group by created_at.
+	SELECT fb.template_version_id, fb.created_at, fb.preset_id, fb.job_status, fb.desired_instances,
+	    ROW_NUMBER() OVER (PARTITION BY fb.preset_id ORDER BY fb.created_at DESC) as rn
+	FROM filtered_builds fb
+),
+failed_count AS (
+    -- Count failed builds per template version/preset in the given period
+	SELECT preset_id, COUNT(*) AS num_failed
+	FROM filtered_builds
+	WHERE job_status = 'failed'::provisioner_job_status
+		AND created_at >= @lookback::timestamptz
+	GROUP BY preset_id
+)
+SELECT tsb.template_version_id,
+	   tsb.preset_id,
+	   COALESCE(fc.num_failed, 0)::int  AS num_failed,
+	   MAX(tsb.created_at)::timestamptz AS last_build_at
+FROM time_sorted_builds tsb
+		 LEFT JOIN failed_count fc ON fc.preset_id = tsb.preset_id
+WHERE tsb.rn <= tsb.desired_instances -- Fetch the last N builds, where N is the number of desired instances; if any fail, we backoff
+  AND tsb.job_status = 'failed'::provisioner_job_status
+  AND created_at >= @lookback::timestamptz
+GROUP BY tsb.template_version_id, tsb.preset_id, fc.num_failed;
```

Comment:
One thing we never check is that the builds we consider are _prebuilds_. That is, regular preset workspace builds that fail or succeed will be counted in computing the backoff for prebuilds. Is that by design?

Intuitively, you'd expect prebuilds to be no different than other preset builds, and have the same failure characteristics. But, I think it's possible that setting a parameter not included in the preset could cause a build to fail. Not a big deal for the odd mistake, but it leaves open an opportunity for malicious DOS of prebuilds by creating a lot of failed builds.

Reply:
good catch, I think we can add a check if it's owned by `prebuilt_user`
@SasSwart wdyt?

Reply:
I fixed it in `filtered_builds` CTE.